File: src/_101/kotlin/Solution.kt
package _101.kotlin

import structure.TreeNode

/**
 * @author relish
 * @since 2018/04/26
 *
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int = 0) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {

    private fun symmetric(left: TreeNode?, right: TreeNode?): Boolean {
        if (left == null && right == null) return true
        if (left == null || right == null) return false
        return left.`val` == right.`val` && symmetric(left.right, right.left) && symmetric(left.left, right.right)
    }

    fun isSymmetric(root: TreeNode?): Boolean {
        if (root == null) return true
        if (root.left == null && root.right == null) return true
        if (root.left == null || root.right == null) return false
        return symmetric(root.left, root.right)
    }
}

fun main(args: Array<String>) {
    println(Solution().isSymmetric(TreeNode.createTestData("[1,2,2,null,3,null,3]")))
}

File: src/_101/Solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    if(root == null) return true
    var x = function(left, right) {
        if(left == null && right == null) return true
        if(left == null || right == null) return false
        if (left.val !== right.val) return false;
        return x(left.left, right.right) && x(left.right, right.left) 
    }
    return x(root.left, root.right)
};